File: app/src/main/java/com/shevaalex/android/rickmortydatabase/ui/location/LocationDetailFragment.java
package com.shevaalex.android.rickmortydatabase.ui.location;

import android.content.Context;
import android.content.res.Configuration;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.shevaalex.android.rickmortydatabase.databinding.FragmentLocationDetailBinding;
import com.shevaalex.android.rickmortydatabase.source.database.CharacterSmall;
import com.shevaalex.android.rickmortydatabase.ui.FragmentToolbarSimple;
import com.shevaalex.android.rickmortydatabase.ui.character.CharacterAuxAdapter;

import java.util.ArrayList;
import java.util.List;

public class LocationDetailFragment extends FragmentToolbarSimple implements CharacterAuxAdapter.OnCharacterListener {
    private FragmentLocationDetailBinding binding;
    private CharacterAuxAdapter characterAuxAdapter;
    private LocationDetailViewModel locationDetailViewModel;
    private List<CharacterSmall> mCharacterList = new ArrayList<>();
    private Context context;

    public LocationDetailFragment() {
        // Required empty public constructor
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        this.context = context;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        locationDetailViewModel = new ViewModelProvider(this).get(LocationDetailViewModel.class);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Set View binding for this fragment;
        binding = FragmentLocationDetailBinding.inflate(inflater, container, false);
        View view = binding.getRoot();
        // retrieve data from parent fragment
        String name = LocationDetailFragmentArgs.fromBundle(requireArguments()).getLocationName();
        String dimension = LocationDetailFragmentArgs.fromBundle(requireArguments()).getLocationDimension();
        String type = LocationDetailFragmentArgs.fromBundle(requireArguments()).getLocationType();
        int locationId = LocationDetailFragmentArgs.fromBundle(requireArguments()).getLocationId();
        //save location id with SavedStateHandle
        locationDetailViewModel.setLocationId(locationId);
        //set retreived data to appropriate views
        if (binding.locationNameValue != null) {
            binding.locationNameValue.setText(name);
        }
        binding.locationDimensionValue.setText(dimension);
        binding.locationTypeValue.setText(type);
        setRecyclerView();
        registerObservers();
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
        if (characterAuxAdapter != null) {
            characterAuxAdapter = null;
        }
    }

    private void setRecyclerView() {
        //set the recyclerview
        LinearLayoutManager layoutManager = new LinearLayoutManager(this.getActivity());
        if (getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE) {
            layoutManager.setOrientation(RecyclerView.HORIZONTAL);
        }
        binding.recyclerviewLocationDetail.setLayoutManager(layoutManager);
        binding.recyclerviewLocationDetail.setHasFixedSize(true);
        //get recyclerview Adapter and set data to it using ViewModel
        characterAuxAdapter = new CharacterAuxAdapter(this, context);
        characterAuxAdapter.setStateRestorationPolicy(RecyclerView.Adapter.StateRestorationPolicy.PREVENT_WHEN_EMPTY);
    }

    private void registerObservers() {
        locationDetailViewModel.getCharactersFromLocation().observe(getViewLifecycleOwner(), characters -> {
            if (!characters.isEmpty()) {
                characterAuxAdapter.setCharacterList(characters);
                mCharacterList = characters;
                binding.recyclerviewLocationDetail.setAdapter(characterAuxAdapter);
            } else {
                binding.locationResidentsNone.setVisibility(View.VISIBLE);
            }
        });
    }

    @Override
    public void onCharacterClick(int position, View v) {
        if (mCharacterList != null && !mCharacterList.isEmpty()) {
            CharacterSmall clickedChar = mCharacterList.get(position);
            LocationDetailFragmentDirections.ToCharacterDetailFragmentAction3 action =
                    LocationDetailFragmentDirections.toCharacterDetailFragmentAction3();
            if (clickedChar != null) {
                action.setCharacterName(clickedChar.getName()).setId(clickedChar.getId());
                Navigation.findNavController(v).navigate(action);
            }
        }
    }
}

File: app/src/main/java/com/shevaalex/android/rickmortydatabase/ui/location/detail/LocationDetailFragment.kt
package com.shevaalex.android.rickmortydatabase.ui.location.detail

import android.content.res.Configuration
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.constraintlayout.motion.widget.MotionLayout
import androidx.constraintlayout.widget.Guideline
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import androidx.navigation.fragment.navArgs
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.shevaalex.android.rickmortydatabase.R
import com.shevaalex.android.rickmortydatabase.RmApplication
import com.shevaalex.android.rickmortydatabase.databinding.FragmentLocationDetailBinding
import com.shevaalex.android.rickmortydatabase.models.character.CharacterModel
import com.shevaalex.android.rickmortydatabase.models.location.LocationModel
import com.shevaalex.android.rickmortydatabase.ui.base.BaseDetailFragment
import com.shevaalex.android.rickmortydatabase.ui.CharacterSmallAdapter
import com.shevaalex.android.rickmortydatabase.utils.Constants
import com.shevaalex.android.rickmortydatabase.utils.DiViewModelFactory
import com.shevaalex.android.rickmortydatabase.utils.safeNavigate
import com.shevaalex.android.rickmortydatabase.utils.setTopPaddingForStatusBar
import javax.inject.Inject

class LocationDetailFragment : BaseDetailFragment<FragmentLocationDetailBinding, LocationModel>() {

    @Inject
    lateinit var viewModelFactory: DiViewModelFactory<LocationDetailViewModel>

    private var adapter: CharacterSmallAdapter? = null

    override val keyDetailObject: String = Constants.KEY_FRAGMENT_LOCATION_DETAIL_OBJECT

    override val viewModel: LocationDetailViewModel by viewModels {
        viewModelFactory
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        getArgs()?.let {
            setupViews(it)
        }
        setRecyclerView()
        registerObservers()
        binding.buttonBack?.let {
            setBackButton(it)
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        //restore the view state
        savedInstanceState?.let {
            restoreViewState(it)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        adapter = null
    }

    /**
     * gets passed data from list fragment
     */
    private fun getArgs(): LocationModel? {
        val args: LocationDetailFragmentArgs by navArgs()
        return args.locationObject
    }

    private fun setupViews(location: LocationModel) {
        setShareButton(location)
        viewModel.setDetailObject(location)
        binding.layoutFragmentLocationDetail.transitionName =
                Constants.TRANSITION_LOCATION.plus(location.id)
        binding.imageLocation?.let {
            setMainImage(
                    imageUrl = location.imageUrl,
                    imageView = it,
                    height = 400,
                    width = 568,
                    placeholderDrawableResource = R.drawable.location_placeholder
            )
        }
        binding.locationName.text = location.name
        binding.locationTypeValue.text = location.type
        binding.locationDimensionValue.text = location.dimension
    }

    private fun setRecyclerView() {
        //set the recyclerview
        val layoutManager = LinearLayoutManager(requireContext())
        if (resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) {
            layoutManager.orientation = RecyclerView.HORIZONTAL
        }
        binding.recyclerviewLocationDetail.layoutManager = layoutManager
        binding.recyclerviewLocationDetail.setHasFixedSize(true)
        //get recyclerview Adapter and set data to it using ViewModel
        adapter = CharacterSmallAdapter(object : CharacterSmallAdapter.CharacterClickListener {
            override fun onCharacterClick(character: CharacterModel) {
                navigateCharacterDetail(character)
            }
        })
        adapter?.stateRestorationPolicy =
                RecyclerView.Adapter.StateRestorationPolicy.PREVENT_WHEN_EMPTY
        binding.recyclerviewLocationDetail.adapter = adapter
    }

    private fun registerObservers() {
        viewModel.characters.observe(viewLifecycleOwner, { characters ->
            characters?.let {
                adapter?.setCharacterList(it)
                //explicilty reset the RV adapter, otherwise RV is invisible at the start of transition or landscape
                binding.recyclerviewLocationDetail.adapter = adapter
                binding.locationResidentsNone.visibility = if (characters.isEmpty()) {
                    View.VISIBLE
                } else View.GONE
            }
        })
    }

    private fun setShareButton(location: LocationModel) {
        binding.buttonShare?.setOnClickListener {
            shareImageWithGlide(location.name, location.imageUrl)
        }
    }

    private fun navigateCharacterDetail(character: CharacterModel) {
        val action = LocationDetailFragmentDirections.actionGlobalCharacterDetailFragment2(character)
        findNavController().safeNavigate<LocationDetailFragment>(action)
    }

    private fun restoreViewState(savedInstanceState: Bundle?) {
        savedInstanceState?.let {
            (it[keyDetailObject] as LocationModel?)?.let { location ->
                viewModel.setDetailObject(location)
            }
        }
    }

    override fun injectFragment() {
        activity?.run {
            (application as RmApplication).appComponent
        }?.inject(this)
    }

    override fun setBinding(inflater: LayoutInflater, container: ViewGroup?) =
            FragmentLocationDetailBinding.inflate(inflater, container, false)

    /**
     * returns MotionLayout. cast needed due to layout being ConstraintLayout in landscape mode
     */
    @Suppress("USELESS_CAST")
    override fun getMotionLayout(): MotionLayout? {
        return try {
            binding.layoutFragmentLocationDetail as MotionLayout
        } catch (e: Exception) {
            null
        }
    }

    override fun getToolbarBottomGuideline(): Guideline? {
        return binding.guidelineToolbar
    }

    override fun getToolbarTopGuideline(): Guideline? {
        return binding.guidelineStatusBar
    }

    override fun setupDetailLayoutWithTransparentStatusBar() {
        binding.layoutFragmentLocationDetail.setTopPaddingForStatusBar()
    }

}